woocommerce/woocommerce-gateway-stripe
# PR: SCA: Add PaymentIntent confirmations as a step in the checkout

Fixes #840 .

#### Changes proposed in this Pull Request:

The original purpose of this PR was to mark orders as failed when an SCA authorization fails. My initial plan was to fire an AJAX call from JS to the server in order to verify that the intent failed, and set the order status to "Payment Failed".

This status was already being updated by webhooks, but only when they are set up, which is not the case for most of our merchants. While considering situations with disabled webhooks and trying to setup the proper AJAX calls, I came to a couple of realizations:

1. The code in the current `sca/master` branch might interfere with extensions, which use the `woocommerce_payment_successful_result` filter, and there seem to be a lot of those: https://github.com/search?q=woocommerce_payment_successful_result&type=Code
2. It is possible that in certain flows merchants would never reach the "Order Received" page, which would prevent the PI/order status from being checked.

In order to deal with those issues while also handling the failed SCA auth scenario, I implemented the following changes in this PR:

1. `WC_Gateway_Stripe::process_payment` no longer returns a hash on the checkout page. Instead, it returns an array containing `result` (`success`), `redirect` (standard Order Received Page), and most importantly an `intent_secret` property, which contains the PI secret. This property is later used within `modify_successful_payment_result` to generate the hash. Since the filter is added with a high priority (99999), it should be able to make use of `redirect` URLs, modified by other extensions.
2. Instead of including the URL of the "Order Received" page in the redirect hash, and redirecting to it on success, there is a new intermediate URL/step.
The new URL looks like this: `https://domain.com/?wc-ajax=wc_stripe_verify_intent&order=1229&nonce=a5fb595c7e&redirect_to=http%3A%2F%2Fdomain.com%2Fcheckout%2Forder-received%2F1229%2F%3Fkey%3Dwc_order_MRBXtOqgSdI93
    - `wc_stripe_verify_intent` is the name of the hook, handled in the new `WC_Stripe_Intent_Controller`. This class is needed because `WC_Gateway_Stripe` is not initialized early enough.
    - The order ID and nonce are used in order to locate the order to check. Based on the order meta, the ID of the PaymentIntent is already accessible and its status can already be checked (`WC_Gateway_Stripe::verify_intent_after_checkout`).
    - `redirect_to` includes the URL, generated during checkout, including all filters applied to `woocommerce_payment_successful_result`.

Our JS will redirect to this new URL in order for the server to check the order/payment status, and only then will the client be redirected to the final Order Received screen.

#### Testing instructions

There is no single result to expect from those changes, so testing flows might differ, but it would be best to check orders' notes after a combination of the following scenarios:

- Start on the checkout page (no order exists yet) __or__ to submit the checkout form, and navigate to the "My Orders" page once the PaymentIntent modal gets shown. This way you can test the "Pay for Order" screen.
- Accept __or__ Reject the SCA authentication.
- Do all of the above with webhooks enabled __or__ disabled.

In all of those cases you should see the proper order notes and status changes appear in the admin.

Note: PI checks are also performed on the "Order Received" screen through this line:

```php
add_filter( 'woocommerce_thankyou_order_id', array( $this, 'thankyou_order_id' ), 100 );
```

Make sure to test the process both when the filter is enabled and disabled.

Question: After all of those redirects, should we keep this filter at all?
